keras_hub/src/utils/transformers/export_gemma_to_safetensor.py

```diff
@@ -0,0 +1,200 @@
+        key_kernel = torch.from_numpy(
+            key_kernel
+        ).T  # Shape: [head_dim, hidden_size]
+        weights_dict[f"model.layers.{i}.self_attn.k_proj.weight"] = key_kernel
+
+        # Attention value projection (v_proj)
+        value_kernel = decoder_layer.attention.value_dense.get_weights()[0][
+            0
+        ]  # Shape: [hidden_size, head_dim]
+        value_kernel = torch.from_numpy(
+            value_kernel
+        ).T  # Shape: [head_dim, hidden_size]
+        weights_dict[f"model.layers.{i}.self_attn.v_proj.weight"] = value_kernel
+
+        # Attention output projection (o_proj)
+        out_kernel = decoder_layer.attention.output_dense.get_weights()[
+            0
+        ]  # Shape: [num_heads, head_dim, hidden_size]
+        out_kernel = (
+            torch.from_numpy(out_kernel)
+            .permute(2, 0, 1)
+            .reshape(backbone.hidden_dim, -1)
+        )  # Reshape to [hidden_size, num_heads * head_dim]
+        weights_dict[f"model.layers.{i}.self_attn.o_proj.weight"] = out_kernel
+
+        # Post-attention normalization (post_attention_layernorm)
+        post_attn_norm = decoder_layer.pre_ffw_norm.get_weights()[
+            0
+        ]  # Shape: [hidden_size]
+        weights_dict[f"model.layers.{i}.post_attention_layernorm.weight"] = (
+            torch.from_numpy(post_attn_norm)
+        )
+
+        # MLP gate projection
+        gate_kernel = decoder_layer.gating_ffw.get_weights()[
+            0
+        ]  # Shape: [hidden_size, intermediate_size]
+        gate_kernel = torch.from_numpy(
+            gate_kernel
+        ).T  # Shape: [intermediate_size, hidden_size]
+        weights_dict[f"model.layers.{i}.mlp.gate_proj.weight"] = gate_kernel
+
+        # MLP up projection
+        up_kernel = decoder_layer.gating_ffw_2.get_weights()[
+            0
+        ]  # Shape: [hidden_size, intermediate_size]
+        up_kernel = torch.from_numpy(
+            up_kernel
+        ).T  # Shape: [intermediate_size, hidden_size]
+        weights_dict[f"model.layers.{i}.mlp.up_proj.weight"] = up_kernel
+
+        # MLP down projection
+        down_kernel = decoder_layer.ffw_linear.get_weights()[
+            0
+        ]  # Shape: [intermediate_size, hidden_size]
+        down_kernel = torch.from_numpy(
+            down_kernel
+        ).T  # Shape: [hidden_size, intermediate_size]
+        weights_dict[f"model.layers.{i}.mlp.down_proj.weight"] = down_kernel
+
+    # Map final normalization
+    final_norm = backbone.get_layer("final_normalization").get_weights()[
+        0
+    ]  # Shape: [hidden_size]
+    weights_dict["model.norm.weight"] = torch.from_numpy(final_norm)
+
+    # Tie lm_head.weight to embedding weights
+    weights_dict["lm_head.weight"] = weights_dict["model.embed_tokens.weight"]
+
+    # Load weights into Hugging Face model
+    print("Loading weights into Hugging Face model...")
+    hf_model.load_state_dict(weights_dict, strict=False)
```

Comment:
This is the line we probably most need to avoid to avoid the double allocation. Can we try writing the weights out with safetensors directly with the safetensors library here and avoiding this?

That will mean we might need to save the config.json separately, outside of `save_pretrained`.

Alternately we could see if we could use the torch meta device to avoid allocating memory, but I don't know if that would work when needing to actually save the model.